### PR TITLE
[MRG] control Hessian term explosion

### LINKED ISSUE #289: loss grows to infinity with cdfast
I have a dataset for which the loss grows without bound using the cdfast solver, but converges normally with batch gradient. Below I've attached the datafiles to reproduce the error in case someone can figure out the problem. Otherwise, I would really appreciate any suggestions on how I can go about debugging this. Thanks!

Here's the problem happening:
```
In [81]: glm = pyglmnet.GLM(verbose=True, solver='cdfast', max_iter=50, alpha=0)

In [82]: glm.fit(bad_exog, bad_endog)
Lambda: 0.1000
Iteration: 0. Loss: 2.510834
Iteration: 1. Loss: 3.258520
Iteration: 2. Loss: 6.065792
Iteration: 3. Loss: 33.602209
Iteration: 4. Loss: 526.502393
Iteration: 5. Loss: 520065.195320
Iteration: 6. Loss: inf
Iteration: 7. Loss: nan
Iteration: 8. Loss: nan
Iteration: 9. Loss: nan
Iteration: 10. Loss: nan
Iteration: 11. Loss: nan
... [Continues until max iter reached]
```

Here's how it works with batch-gradient:
```
In [83]: glm = pyglmnet.GLM(verbose=True, solver='batch-gradient', max_iter=50, alpha=0)

In [84]: glm.fit(bad_exog, bad_endog)
Lambda: 0.1000
Iteration: 0. Loss: 2.527621
Iteration: 1. Loss: 1.852870
Iteration: 2. Loss: 1.111185
Iteration: 3. Loss: 0.518860
Iteration: 4. Loss: 0.196907
Iteration: 5. Loss: 0.071565
Iteration: 6. Loss: 0.033076
Iteration: 7. Loss: 0.022561
Iteration: 8. Loss: 0.018103
Iteration: 9. Loss: 0.015213
Iteration: 10. Loss: 0.012827
Iteration: 11. Loss: 0.010914
Iteration: 12. Loss: 0.009259
Iteration: 13. Loss: 0.007895
Iteration: 14. Loss: 0.006712
Iteration: 15. Loss: 0.005725
Iteration: 16. Loss: 0.004872
Iteration: 17. Loss: 0.004156
Iteration: 18. Loss: 0.003538
Iteration: 19. Loss: 0.003017
Iteration: 20. Loss: 0.002570
Iteration: 21. Loss: 0.002191
Iteration: 22. Loss: 0.001867
Iteration: 23. Loss: 0.001591
Iteration: 24. Loss: 0.001356
Iteration: 25. Loss: 0.001156
	Converged in 26 iterations
```

Here are the data `bad_endog` and `bad_exog` for reproducing the problem.
https://www.dropbox.com/s/v3llq1umhrfxu49/bad_endog?dl=1
https://www.dropbox.com/s/vf0bhko9auguf7p/bad_exog?dl=1
They are plaintext and can be loaded as follows:
```
bad_endog2 = np.loadtxt('bad_endog', dtype=np.int)
bad_exog2 = np.loadtxt('bad_exog', dtype=np.float)
```

### PR DESCRIPTION
**Pull request by pavanramkumar**
closes #289

### CODE REVIEW

File: pyglmnet/pyglmnet.py [hunk 1]
```diff
@@ -730,8 +730,13 @@ def _cdfast(self, X, y, ActiveSet, beta, rl, fit_intercept=True):
                 gk += reg_scale * gk_reg
                 hk += reg_scale * hk_reg
 
+                # Ensure that update does not blow up if Hessian is small
+                if hk < 1e-3:
```
Comment (pavanramkumar) [inline pyglmnet/pyglmnet.py]: make `1e-3` into a parameter with defaults

**Comment by pavanramkumar** [comment]
I think the intuition of @cxrodgers is right that close to the true solution the Hessian term is small and therefore the update term explodes. I've put in a PR #348 that only uses the Hessian reciprocal if it is sufficiently large and added a test to ensure the estimated coefficients are not `np.nan`

@cxrodgers if you have time, it will be great to test on your dataset to check if it alleviates your convergence problem before we merge this

File: pyglmnet/pyglmnet.py (hunk 1 above)
Comment (pavanramkumar) [reply pyglmnet/pyglmnet.py]: done

**Comment by pavanramkumar** [comment]
@geektoni this PR may also address the nan issue for cdfast in #274

File: pyglmnet/pyglmnet.py [hunk 2]
```diff
@@ -472,12 +472,17 @@ class GLM(BaseEstimator):
...
     tol: float
         convergence threshold or stopping criteria.
-        Optimization loop will stop when norm(gradient) is below the threshold.
+        Optimization loop will stop when relative change
+        in parameter norm is below the threshold.
         default: 1e-6
+    htol: float
+        tolerance for Hessian term. if the Hessian is below this value
+        the learning rate will be used as step size.
+        default: 1e-3
```
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: as we discussed privately, we don't need to expose it for now. If for some reason we realize that users need to tweak it, we can always expose it later. However, "unexposing" is harder because you will need to have a deprecation cycle then. We can have the `htol` declared as a global variable so that it's not a dangling constant in the code.

Comment (pavanramkumar) [reply pyglmnet/pyglmnet.py]: i've simplified it even further. we simply do `1. / (1. + hk)` instead of `1. / hk`: it's a special case of Quasi Newton for 1 dimensional Newton update for a convex problem

**Comment by pavanramkumar** [comment]
@jasmainak @cxrodgers 

i have also now tested with the dataset provided in #289 

i verified that both `cdfast` and `batch-gradient` converge with smoothly decreasing loss trace across `poisson`, `softplus` and `gaussian` distrs.

**Comment by jasmainak** [comment]
@pavanramkumar can you update `whats_new.rst` ?

File: pyglmnet/pyglmnet.py [hunk 3]
```diff
@@ -730,8 +732,10 @@ def _cdfast(self, X, y, ActiveSet, beta, rl, fit_intercept=True):
                 gk += reg_scale * gk_reg
                 hk += reg_scale * hk_reg
 
+                # Ensure that update does not blow up if Hessian is small
+                update = 1. / (1 + hk) * gk
```
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: It would be nice if we could also add a sentence in the tutorial

**Comment by pavanramkumar** [comment]
@jasmainak updated `whats_new.rst` and `tutorial.rst`

File: doc/whats_new.rst [hunk 4]
```diff
@@ -14,6 +14,8 @@ Current
 Changelog
 ~~~~~~~~~
 
+    - Graceful handling of small Hessian term in coordinate descent solver by `Pavan Ramkumar`_.
```
Comment (jasmainak) [inline doc/whats_new.rst]: This would go in the bug section, no?
```suggestion
    - Graceful handling of Hessian term that exploded when small, in coordinate descent solver by `Pavan Ramkumar`_.
```

File: pyglmnet/pyglmnet.py [hunk 5]
```diff
@@ -730,8 +732,10 @@ def _cdfast(self, X, y, ActiveSet, beta, rl, fit_intercept=True):
                 gk += reg_scale * gk_reg
                 hk += reg_scale * hk_reg
 
+                # Ensure that update does not blow up if Hessian is small
+                update = 1. / hk * gk if hk > 1. else self.learning_rate * gk
```
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: why not `abs(hk) > 1.`?

Comment (pavanramkumar) [reply pyglmnet/pyglmnet.py]: `hk` is positive for convex objective

File: doc/whats_new.rst (hunk 4 above)
Comment (pavanramkumar) [reply doc/whats_new.rst]: done

**Comment by jasmainak** [comment]
Merged, thanks @pavanramkumar !
